torch/_refs/__init__.py

```diff
@@ -4328,20 +4328,21 @@ def linspace(
     if start == end:
         return torch.full((steps,), start, dtype=dtype, **factory_kwargs)  # type: ignore[arg-type]
 
-    # arange returns values in the interval [start, end) so we add an an eps to make it [start, end]
-    # The eps is small enough as to always add just the element end
-    step_size = 1 / (steps - 1)
-    eps = step_size / 2
-    # arange returns a tensor of size divup(end - start, step) and thus, for the arguemnts below
-    # ceil(div(1 + step_size/2,  1/(steps - 1)) = steps - 1  + ceil(1 / 2) = steps
-    # torch.arange is an scan algorithm, so we need a high-precision dtype
-    rg = torch.arange(
-        0, 1 + eps, step_size, dtype=torch.float64, **factory_kwargs  # type: ignore[arg-type]
+    # Perform in arange in int for exactness, so that out[0] == start, out[-1] == end
+    rg = torch.arange(0, steps, **factory_kwargs)  # type: ignore[arg-type]
+
+    # Small types need to be computed in higher precision as this is, at heart, an associative scan
+    dtype_red = (
+        torch.int64
+        if (utils.is_boolean_dtype(dtype) or utils.is_integer_dtype(dtype))
+        else dtype
     )
-    double_dtype = torch.complex128 if utils.is_complex_dtype(dtype) else torch.float64
-    rg = _maybe_convert_to_dtype(rg, double_dtype)  # type: ignore[assignment]
-    cast = partial(torch.full, (1,), dtype=double_dtype, **factory_kwargs)
-    out = torch.lerp(cast(start), cast(end), rg)
+    computation_dtype, _ = utils.reduction_dtypes(
+        rg, REDUCTION_OUTPUT_TYPE_KIND.SAME, dtype_red
+    )
+
+    cast = partial(torch.full, (), dtype=computation_dtype, **factory_kwargs)
+    out = rg * cast((end - start) / (steps - 1)) + cast(start)
```

Comment:
Should this not use the algorithm from ATen:

```
where(ind < halfway, start + step * rg, end - step * (steps - rg - 1))
```

This should get the end points of the range exactly.

Reply:
Sure thing. Let's kill two birds with one stone.

Reply:
https://github.com/pytorch/pytorch/pull/91625

Reply:
Why do we need to cast `end` and `start` to tensor? Inductor remedies this, but it still looks unnecessary?

Reply:
Sure, let me cast `rg` instead.